### ISSUE TITLE
Switch to branchless binary search 

### ISSUE THREAD

**Issue by lemire** [lemire/RoaringBitmap#48]
We use a conventional binary search for our array containers. In a random microbenchmark, we get that a branchless version is significantly faster.

      ShortBinarySearch.branchlessBinarySearch2     32  avgt    5   12194.481 ±   13.544  ns/op
      ShortBinarySearch.branchyBinarySearch         32  avgt    5   16102.804 ±  245.875  ns/op
      
      ShortBinarySearch.branchlessBinarySearch2    128  avgt    5   16681.683 ±   23.848  ns/op
      ShortBinarySearch.branchyBinarySearch        128  avgt    5   24971.793 ±  116.461  ns/op
      
      ShortBinarySearch.branchlessBinarySearch2   1024  avgt    5   26024.953 ±   31.438  ns/op
      ShortBinarySearch.branchyBinarySearch       1024  avgt    5   43340.094 ±   98.262  ns/op
      
      ShortBinarySearch.branchlessBinarySearch2   2048  avgt    5   31097.226 ±  167.894  ns/op
      ShortBinarySearch.branchyBinarySearch       2048  avgt    5   47803.827 ± 3093.436  ns/op

As you can see, the gain can be large. It is not 2x, but nearly so.

Here is the new function that could be put in the code:

https://github.com/lemire/microbenchmarks/blob/master/src/main/java/me/lemire/microbenchmarks/binarysearch/ShortBinarySearch.java#L118-L135

*However*, we should not trust the microbenchmark.  One should check whether it does improve performance genuinely in cases we care about.

**Comment by lemire** [comment]
I have implemented branchless binary search in a branch of the project (https://github.com/lemire/RoaringBitmap/tree/branchlessbinsearch).

The results are inconclusive. There are large gains and large losses, depending on the dataset. It is unclear why but a possibility is that the branchless approach suffers from unpredictable array sizes. My benchmark is *not* written in a way where we repeatedly query the same arrays as I do not expect users of this library to proceed in such a manner.

Someone should review my code changes and the benchmark. I might have missed something, of failed to take advantage of an obvious optimization. (c.c. @gssiyankai ).


```
$ cd jmh
$ ./run.sh needwork.RandomAccess

Benchmark                                           (dataset)  Mode  Cnt       Score      Error  Units
RandomAccess.branchlessRoaring                  census-income  avgt    5   11434.537 ±  116.724  us/op
RandomAccess.branchlessRoaring                     census1881  avgt    5    4103.713 ±   66.188  us/op
RandomAccess.branchlessRoaring                  dimension_008  avgt    5   55197.550 ± 1015.186  us/op
RandomAccess.branchlessRoaring                  dimension_003  avgt    5  158696.314 ± 2648.847  us/op
RandomAccess.branchlessRoaring                  dimension_033  avgt    5    3865.845 ±    5.959  us/op
RandomAccess.branchlessRoaring                   uscensus2000  avgt    5    2014.063 ±   41.530  us/op
RandomAccess.branchlessRoaring                weather_sept_85  avgt    5   14627.764 ±  698.406  us/op
RandomAccess.branchlessRoaring             wikileaks-noquotes  avgt    5    6997.703 ±   25.780  us/op
RandomAccess.branchlessRoaring              census-income_srt  avgt    5    9977.109 ±  103.355  us/op
RandomAccess.branchlessRoaring                 census1881_srt  avgt    5    4322.006 ±    2.309  us/op
RandomAccess.branchlessRoaring            weather_sept_85_srt  avgt    5   10095.089 ±  170.911  us/op
RandomAccess.branchlessRoaring         wikileaks-noquotes_srt  avgt    5    5566.580 ±    5.131  us/op
RandomAccess.branchlessRoaringWithRun           census-income  avgt    5   13265.026 ±   60.324  us/op
RandomAccess.branchlessRoaringWithRun              census1881  avgt    5    3542.456 ±   38.696  us/op
RandomAccess.branchlessRoaringWithRun           dimension_008  avgt    5   67959.608 ±  559.445  us/op
RandomAccess.branchlessRoaringWithRun           dimension_003  avgt    5  143621.358 ±  365.676  us/op
RandomAccess.branchlessRoaringWithRun           dimension_033  avgt    5    3469.598 ±    4.465  us/op
RandomAccess.branchlessRoaringWithRun            uscensus2000  avgt    5    2449.567 ±  150.991  us/op
RandomAccess.branchlessRoaringWithRun         weather_sept_85  avgt    5   16331.381 ±  731.436  us/op
RandomAccess.branchlessRoaringWithRun      wikileaks-noquotes  avgt    5    6819.558 ±   44.194  us/op
RandomAccess.branchlessRoaringWithRun       census-income_srt  avgt    5   10749.394 ±   99.160  us/op
RandomAccess.branchlessRoaringWithRun          census1881_srt  avgt    5    5150.419 ±    4.540  us/op
RandomAccess.branchlessRoaringWithRun     weather_sept_85_srt  avgt    5    9869.339 ±   12.471  us/op
RandomAccess.branchlessRoaringWithRun  wikileaks-noquotes_srt  avgt    5    6435.810 ±   34.548  us/op
RandomAccess.branchyRoaring                     census-income  avgt    5   11611.748 ±  141.998  us/op
RandomAccess.branchyRoaring                        census1881  avgt    5    4714.165 ±  177.314  us/op
RandomAccess.branchyRoaring                     dimension_008  avgt    5   52888.820 ±   85.368  us/op
RandomAccess.branchyRoaring                     dimension_003  avgt    5  156059.006 ±  766.946  us/op
RandomAccess.branchyRoaring                     dimension_033  avgt    5    4739.661 ±  186.108  us/op
RandomAccess.branchyRoaring                      uscensus2000  avgt    5    3137.948 ±   52.843  us/op
RandomAccess.branchyRoaring                   weather_sept_85  avgt    5   13403.414 ±  571.237  us/op
RandomAccess.branchyRoaring                wikileaks-noquotes  avgt    5    7426.687 ±   47.852  us/op
RandomAccess.branchyRoaring                 census-income_srt  avgt    5   11587.898 ±  192.717  us/op
RandomAccess.branchyRoaring                    census1881_srt  avgt    5    4710.917 ±  117.467  us/op
RandomAccess.branchyRoaring               weather_sept_85_srt  avgt    5   10732.727 ±  345.394  us/op
RandomAccess.branchyRoaring            wikileaks-noquotes_srt  avgt    5    5779.323 ±   71.697  us/op
RandomAccess.branchyRoaringWithRun              census-income  avgt    5   12381.902 ±  120.366  us/op
RandomAccess.branchyRoaringWithRun                 census1881  avgt    5    4637.310 ±  106.824  us/op
RandomAccess.branchyRoaringWithRun              dimension_008  avgt    5   50070.772 ±   25.785  us/op
RandomAccess.branchyRoaringWithRun              dimension_003  avgt    5  136020.216 ±  325.160  us/op
RandomAccess.branchyRoaringWithRun              dimension_033  avgt    5    3857.788 ±  224.316  us/op
RandomAccess.branchyRoaringWithRun               uscensus2000  avgt    5    3018.201 ±   29.130  us/op
RandomAccess.branchyRoaringWithRun            weather_sept_85  avgt    5   13851.379 ±  381.538  us/op
RandomAccess.branchyRoaringWithRun         wikileaks-noquotes  avgt    5    6777.768 ±   33.520  us/op
RandomAccess.branchyRoaringWithRun          census-income_srt  avgt    5   11547.330 ±   10.219  us/op
RandomAccess.branchyRoaringWithRun             census1881_srt  avgt    5    4490.518 ±   39.298  us/op
RandomAccess.branchyRoaringWithRun        weather_sept_85_srt  avgt    5   10040.633 ±   67.540  us/op
RandomAccess.branchyRoaringWithRun     wikileaks-noquotes_srt  avgt    5    5149.222 ±   33.012  us/op
```

Reference: https://github.com/lemire/RoaringBitmap/blob/branchlessbinsearch/jmh/src/main/java/org/roaringbitmap/needwork/RandomAccess.java

**Comment by gssiyankai** [comment]
@lemire ,
I will have a look soon.

**Comment by lemire** [comment]
+1

**Comment by lemire** [comment]
I have updated the benchmark so that there is more data than fits in CPU cache. That is, we still have small arrays, we have many of them. Branchless binary  search then becomes awful:

```
Benchmark                                  (N)   Mode  Cnt   Score   Error  Units
ShortBinarySearch.branchlessBinarySearch  1024  thrpt    5   3.506 ± 0.118  ops/s
ShortBinarySearch.branchyBinarySearch     1024  thrpt    5  15.608 ± 4.454  ops/s
```

There might be a way to "prefetch" the data somehow... but I don't know how to do it.

**Comment by lemire** [comment]
I might have found an approach that is faster than the conventional binary search:

https://github.com/lemire/microbenchmarks/blob/master/src/main/java/me/lemire/microbenchmarks/binarysearch/ShortBinarySearch.java#L128-L167

We simply unroll the loop which makes it easier for the processor to prefetch the right data.

**Comment by lemire** [comment]
Version 0.5.7 (to be released today) has faster binary search. Closing.

**Comment by gssiyankai** [comment]
Hybrid binary search in the buffer package is active in version 0.5.7 (and
also in version 0.5.8 if I'm not mistaken)

**Comment by lemire** [comment]
@gssiyankai 

I got confused somehow. I need to release 0.5.9. Thanks.

**Comment by lemire** [comment]
@gssiyankai  Hopefully, this has been fixed by release 0.5.9.

**Comment by gssiyankai** [comment]
@lemire  :+1:
